### ISSUE TITLE
Debounce URL updates

### ISSUE THREAD

**Issue by nick-keller** [47ng/nuqs#291]
Hello,
For now, it cannot be used to save the state of a search input because the URL is updated for each key stroke. It would be nice to debounce or throttle the URL updates for a smoother interaction ;)

**Comment by franky47** [comment]
I would argue that this is not the scope of this library, there are hooks that allow debouncing and throttling methods:
- [`useDebounce`](https://github.com/streamich/react-use/blob/master/docs/useDebounce.md)
- [`useThrottle`](https://github.com/streamich/react-use/blob/master/docs/useThrottle.md)

**Comment by franky47** [comment]
FYI, this was implemented in 1.8+, since changes to the history API are rate-limited by the browser, throttling was actually necessary for all updates.

**Comment by franky47** [comment]
@rwieruch one issue with denouncing the call to set state is, like React.useState, you'd also delay the internal state representation. If using a controlled `<input value={state} ...`, that will lead to the input contents lagging behind and skipping data if typing too fast.

`nuqs` uses an internal React state that can be connected to high-frequency inputs (like text inputs or sliders), and only throttles the updates to the URL. Combined with `shallow: false`, this lets the user control how frequently the server is sent the new state URL.

https://nuqs.47ng.com/docs/options#throttling-url-updates

**Comment by rwieruch** [comment]
I see, thanks for the nudge in the right direction! Would it make sense to have a `debounceMs` too? Somehow I am used to debounce such requests rather than throttling them 😅 Throttle works too though.

**Comment by Winwardo** [comment]
Just came across this - a built in `debounceMs` would be wonderful. Say I wish to debounce a search by 500ms. Using `throttleMs` will mean the first character input will be sent to search, and then the rest of the typing will be throttled, which is an odd user experience.

I see that there are other ways to achieve it (e.g. `use-debounce`) but if throttle is already implemented, it feels like debounce is a natural pairing.

(Thanks for the great library by the way!)

**Comment by shane-downes** [comment]
+1 for `debounceMs`

I'm managing some query filters and a debounced search box using router query state. I've started refactoring this part of the app to make it more maintainable as we expand the search features. 

I had actually thought of writing a state/router wrapper with a similar API to `useQueryState` for internal use, but found `nuqs` today and it seems like a great fit (awesome work btw, thank you!).

On initial testing the package works brilliantly for this - except that using `throttleMs` leads to strange UX as [Winwardo](https://github.com/Winwardo) noted [above](https://github.com/47ng/nuqs/issues/291#issuecomment-1979224219). 

I think `debounceMs` would be a great fit for the package because of how well throttle almost works out of the box for this use case.

Potential example with `debounceMs` (Pages router / RTKQ):

```typescript
const searchQuery = getQueryString(router.query.search);

const [search, setSearch] = useQueryState('search', { debounceMs: 500 });

const { data = serverData } = useGetDataQuery({ search: searchQuery }); // RTKQ

//...

<input value={search} onChange={(e) => setSearch(e.target.value)} />
```

Example without `debounceMs`:

```typescript
const [search, setSearch] = useQueryState('search');

// Have to manually control query now
const [getData, { data = serverData }] = useLazyGetDataQuery();

useEffect(() => {
  const debounce = setTimeout(() => {
    getData({ search });
  }, 500);

  return () => {
    clearTimeout(debounce);
  };
}, [getData, searchQuery]);

//...

<input value={search} onChange={(e) => setSearch(e.target.value)} />
```

**Comment by franky47** [comment]
I see two issues with supporting both throttling and debouncing.

The first is the API: providing both `throttleMs` and `debounceMs` would allow setting both, which results in undefined behaviour. This could be solved by changing the API to something like this (feedback welcome):

```ts
.withOptions({
  limitUrlUpdates: { // better be explicit in what this does
    method: 'throttle' | 'debounce',
    timeMs: 500
  }
})

// Or with helpers:
.withOptions({
  limitUrlUpdates: debounce(500),
  limitUrlUpdates: throttle(500)
})
```

The second point is closely related: both throttle and debounce methods will actually run in parallel, for different hook setups to work together. Example:
```ts
const [, setFoo] = useQueryState('foo', { limitUrlUpdates: debounce(500) })
const [, setBar] = useQueryState('bar', { limitUrlUpdates: throttle(500) })

const doubleUpdate = () => {
  setFoo('foo') // This will update immediately
  setBar('bar') // This will update in 500ms
}
```
I'll see what I can do to refactor the URL update queue system to account for both methods.

**Comment by rwieruch** [comment]
Hm. Will the next version be a breaking change anyway? One could consider only supporting debounce and not throttle anymore. But one would have to get some user data here whether debounce is more widely used for URL state.

**Comment by greghart** [comment]
> @rwieruch one issue with debouncing the call to setState is, like React.useState, you'd also delay the internal state representation. If using a controlled `<input value={state} ...`, that will lead to the input contents lagging behind and skipping data if typing too fast.
> 
> `nuqs` uses an internal React state that can be connected to high-frequency inputs (like text inputs or sliders), which updates immediately, and only throttles the updates to the URL. Combined with `shallow: false`, this lets the user control how frequently the server is sent the new state URL.
> 
> https://nuqs.47ng.com/docs/options#throttling-url-updates

I appreciate this getting looked at!

In the mean time, you can still get debouncing with high-frequency inputs by using `useDebounce` (or `lodash/debounce` if you like the leading/trailing options) alongside `useState` in a little hook wrapper like so:
(note I didn't type all the overloads since I don't use them so YMMV)

```
function useQueryStateDebounced<T>(
  key: string,
  options: UseQueryStateOptions<T> & {
    defaultValue: T;
  },
  debounceMs = 350
): UseQueryStateReturn<
  NonNullable<ReturnType<typeof options.parse>>,
  typeof options.defaultValue
> {
  const [valueQs, setQs] = nuqsUseQueryState<T>(key, options);
  const [valueReact, setReact] = React.useState<T | null>(valueQs);
  const debouncedSetQs = React.useCallback(debounce(setQs, debounceMs), [
    setQs,
  ]);
  const set = (newValue: any) => {
    setReact(newValue);
    debouncedSetQs(newValue);
  };
  return [valueReact as typeof valueQs, set as typeof setQs];
}
```

**Comment by floatrx** [comment]
+1 for debounce

**Comment by hauptrolle** [comment]
+1 for debounce. Would be super helpful for a search input, to not hit the server with every keystroke :) 

Btw: Awesome library! ✌️

**Comment by franky47** [comment]
To avoid sending on every keystroke, you can use the `throttleMs` option, but yeah doing an eventually consistent send with the whole query after a certain time of inactivity would be preferable.

Now that v2 is out, debounce support is back on the roadmap!

**Comment by tylersayshi** [comment]
@TkDodo talked in his react advance talk yesterday a little about why react-query chose to not implement debouncing in the core library itself: [youtube link](https://www.youtube.com/live/FAkFJ7Mp6Jk?si=fX_qkzehxISNIw1b&t=15925)

The key points from what I could tell:
- There are many ways to implement debounce and they're not necessarily generalize-able
  - `useDeferredValue` may work for many use cases
- The average consumer of the library may not need to debounce
- It adds to bundle size for the average consumers who won't use it

I don't personally have strong opinions one way or the other, but react-query's experience with `debounce` as a potential feature seems very relevant here.

**Comment by franky47** [comment]
Thanks for the link @tylersayshi !

One key difference here I think, is that in the example given for RQ, the thing to rate-limit is truly external to the library: it's an input passed to the query key and to other options. In this case, composition is indeed a much better API.

The difference in nuqs is that the thing to rate-limit is not exposed to the user: we're not talking about rate-limiting updates of the state value returned by the hook (this can, and should, be done in userland like for RQ), but rate-limiting the sync mechanism that writes to the URL, because of limits imposed by browser vendors.

Since those updates occur outside of the render tree, traditional composition (via hooks) doesn't apply. We could allow some sort of callback mechanism to define custom methods of rate-limiting, but those likely would be less optimised and would not handle certain cases (eg: setting multiple query states in the same tick batches them together into a single URL update).

**Comment by tylersayshi** [comment]
Thanks for the clarification! That makes sense to me :)

**Comment by franky47** [comment]
I started writing some notes about how this would all work internally, and while it would be possible to introduce the `limitUrlUpdates` option incrementally (without a breaking change), there is one feature that might break as a result: the Promise that is returned by the state updater function.

I wrote about this [on BlueSky](https://bsky.app/profile/francoisbest.com/post/3lhjif6jk6222), and would love your thoughts on this matter. I doubt it's a very used feature, but it might break (in behaviour) nonetheless.

TL;DR: currently, multiple state updates (of any keys) pushed in the same event loop tick are guaranteed to return the same Promise reference, that resolves when the URL is next updated (with the URLSearchParams that were applied).

This may no longer be the case: each state update might return a Promise that resolves when _that key_ gets updated in the URL next. It might be possible to make it so that the default behaviour (not specifying any URL rate limiting values) would result in stable Promise references being returned, I'll have to dig deeper.

The architecture that is likely to be implemented is a cascade: one debouncing queue per-key (those will definitely get different Promise references, but it's a new feature so it's alright), possibly throttling queues per-key too (for those that define a custom throttling value, TBC), all feeding into one last global throttling queue that protects against browser rate limits.

**Comment by franky47** [comment]
Debounce support has been achieved internally 🥳

https://github.com/user-attachments/assets/a8496a26-796e-4ab2-9abf-19089ddf24e2

**Comment by franky47** [comment]
@floatrx have you tried #900? I'm looking for folks to try it before I can stabilise the API.

**Comment by floatrx** [comment]
> [@floatrx](https://github.com/floatrx) have you tried [#900](https://github.com/47ng/nuqs/pull/900)? I'm looking for folks to try it before I can stabilise the API.


Not yet, but I'll try and give you feedback later. Thanks!

**Comment by issam-seghir** [comment]
this is what actually worked for me , using `useTransition `with `limitUrlUpdates`

- client component 
```tsx
"use client";
import { Input } from "@/components/ui/input";
import { cn } from "@/lib/utils";
import { CircleX, Search as SearchShadcn } from "lucide-react";
import { useQueryStates } from "nuqs";
import { searchParamsParser } from "@/lib/query-options";
import { useState, useEffect, useTransition, type ChangeEvent } from "react";
import { useDebouncedCallback } from "use-debounce";

export function Search({
    placeholder = "Search ...",
}: {
    placeholder?: string;
}) {
    const [isPending, startTransition] = useTransition();

    const [params, setParams] = useQueryStates(searchParamsParser, {
        shallow: false,
    });

    const [inputValue, setInputValue] = useState(params?.search || "");

    // Sync URL param changes back to input (for browser navigation)
    useEffect(() => {
        if (params?.search !== undefined) {
            setInputValue(params.search);
        }
    }, [params?.search]);

    const debouncedSearch = useDebouncedCallback((value: string) => {
        startTransition(() => {
            setParams(
                {
                    search: value,
                },
                {
                    limitUrlUpdates: {
                        method: value === "" ? "throttle" : "debounce",
                        timeMs: value === "" ? 50 : 800,
                    },
                }
            );
        });
    }, 800);

    const handleSearch = (event: ChangeEvent<HTMLInputElement>) => {
        const value = event.target.value;

        setInputValue(value);
        debouncedSearch(value);
    };

    const clearSearch = () => {
        setInputValue("");

        // Cancel any pending debounced calls
        debouncedSearch.cancel();

        startTransition(() => {
            setParams(
                {
                    search: "",
                },
                {
                    limitUrlUpdates: {
                        method: "throttle",
                        timeMs: 50,
                    },
                }
            );
        });
    };

    return (
        <div className="relative ml-auto flex-1 md:grow-0">
            <SearchShadcn className="absolute right-2.5 top-4 h-[22px] w-[22px] text-muted-foreground ltr:left-2.5" />
            <CircleX
                className={cn(
                    "absolute top-4 h-[22px] w-[22px] cursor-pointer text-muted-foreground opacity-0 transition-all duration-200 hover:stroke-pink-500 ltr:right-2.5 rtl:left-2.5",
                    inputValue && "opacity-100",
                )}
                onClick={clearSearch}
            />
            <Input
                type="search"
                placeholder={placeholder}
                value={inputValue}
                onChange={handleSearch}
                className={cn(
                    "border-[3px] border-solid border-[#eeacc8] bg-white px-8 py-6 ps-[2.5rem] text-lg font-medium focus-visible:ring-0",
                    isPending && "opacity-80" 
                )}
            />
            {isPending && (
                <div className="absolute end-10 top-5 h-[22px] w-[22px]">
                    <div className="h-4 w-4 animate-spin rounded-full border-2 border-gray-300 border-t-pink-500"></div>
                </div>
            )}
        </div>
    );
}
```

**Comment by franky47** [comment]
@issam-seghir a couple of comments & questions on your implementation:

- You don't need an optional chaining on `params?.search`, the state returned by useQueryStates is guaranteed to be an object (but its properties may be nullable).
- You may want to use `defaultRateLimit` rather than `throttle + 50ms`, as it handles Safari's different (higher) rate limits.
- I understand the idea of having the startTransition wrap the state update, but why throttle it as well? In your case you'd end up with cascading throttles of 800ms, resulting in a minimum 1.6s delay between hitting a key and it being reflected in the URL.
- If you want your transition to span the whole duration of the debounce period (ie start on the first update, and resolve when the URL updates), you could drop the curly brackets and return the state updater function's returned Promise to startTransition, so it can clear isPending when it resolves (when the URL is updated with those changes):

```ts
const [isPending, startTransition] = useTransition()
const [, setSearchParams] = useQueryStates(...)

// Only transitions during the internal state update (optimistic UI)
startTransition(() => {
  setSearchParams({ ... })
  // Returns void
})

// Transitions during the whole debounce period
startTransition(() => setSearchParams({ ... }) /* returns a Promise */)
```

**Comment by raviships** [comment]
@franky47 I've tried the #900 , but unfortunately I couldn't figure out a good way to use it in my case.
It might work well with server components. But if we are fetching things on the client side with react-query we need 2 states.
 
1. The immediately updated state to show in the input
2. The debounced state to use in the query

Right now, even though the url is debounced, the actual state is updated immediately, so I can't use it in the query. I could use useSearchParams, but then I feel it would take away a big part of the reason of using nuqs in the first place.

Personally, I would prefer that the nuqs state was also debounced, since I can easily manually create a state to show in the input.

Right now my workaround is to use the "use-debounce" package together with nuqs.

Let me know if I'm missing something or if you need more clarification on the issue.
Also, thanks for the great library!

**Comment by karanuday** [comment]
@perfectbase this is pretty much where Im stuck at right now. But even if you use the workaround, handling back/front navigation to make sure UI component (input) states are in sync with the update url gets quite tricky, especially in a NextJS app due to their internal route caching mechanisms.
